File: 1667-find-kth-bit-in-nth-binary-string/find-kth-bit-in-nth-binary-string.py
class Solution:
    
    def findKthBit(self, n: int, k: int) -> str:
        invert = False
        length = 2**n - 1

        def helper(length, k, invert):
            if length == 1:
                return '0' if not invert  else '1'
            half = length // 2
            if k <= half:
                return helper(half, k, invert)
            elif k > half + 1:
                return helper(half, 1 + length - k, not invert)
            else:
                return '1' if not invert  else '0'

        return helper(length, k, invert)

        # cache = {}
        # def inverse(strs):
        #     result = ""
        #     for s in strs:
        #         if s == '0':
        #             result += '1'
        #         else:
        #             result += '0'
        #     return result

        # def dfs(i):
        #     if i in cache: return cache[i]
        #     if i == 1:
        #         return '0'

        #     cache[i] = dfs(i-1) + "1" + inverse(dfs(i-1))[::-1]
        #     return cache[i]

        # return dfs(n)[k - 1]

File: 1667-find-kth-bit-in-nth-binary-string/find-kth-bit-in-nth-binary-string.cpp
class Solution {
public:
    // recursion with variable
    // char recurssion(int length, int k){
    //     if (length == 1) return '0';
    //     int half = length / 2;
    //     if (k <= half){
    //         return recurssion(half, k);
    //     }
    //     else if (k > half + 1){
    //         int res = recurssion(half, 1 + length - k );
    //         return res == '1' ? '0' : '1';
    //     }
    //     else{
    //         return '1';
    //     }
    // }

    // recurssion without variable
    // char recurssion(int length, int k, bool invert){
    //     if (length == 1) return invert == false ? '0' : '1';
    //     int half = length / 2;
    //     if (k <= half) return recurssion(half, k, invert);
    //     else if (k > half + 1) return recurssion(half, 1 + length - k, !invert);
    //     else return invert == false ? '1' : '0';
    // }


    char findKthBit(int n, int k) {
        int length = pow(2, n) - 1;
        // return recurssion(length, k, false);
        // iterative
        bool invert = false;
        while (length > 1){
            int half = length / 2;
            if (k <= half) length = half;
            else if (k > half + 1){
                k = 1 + length - k;
                length = half;
                invert = !invert;
            }
            else return invert == false ? '1' : '0';
        }
        return invert == false ? '0' : '1';
    }
};